### ISSUE TITLE
ActiveQuery needs AFTER_FIND event

### ISSUE THREAD

**Issue by tanakahisateru** [yiisoft/yii2#10427]
I want to operate models just after fetching for individual queries. But sometimes ActiveQuery would be evaluated lazily.

I propose to trigger `ActiveQuery::EVENT_AFTER_FIND` event against to ActiveQuery's listener.

``` php
$dataProvider->query = $foo->getRelation('bars');
$dataProvider->query->on(ActiveQuery::EVENT_AFTER_FIND, function($event) use($foo) {
    foreach ($event->models as $bar) {
        // Manual relation setup needed because:
        // - inverseOf() is ignored because this is not a property access.
        // - To use eager loading ->with('foo') creates another instance from $foo.
        $bar->populateRelation('foo', $foo);

        // Users can do any other preparations. ActiveRecord::EVENT_AFTER_FIND is
        // too common and it can't use contextual variables of query time.
    }
});
```

Currently I should do:

``` php
<div class="bars-list">
    <?= \yii\widgets\ListView::widget([
        'dataProvider' => $dataProvider,
        'itemView' => function($bar) use($foo) {
            /** @var ActiveRecord $bar */
            $bar->populateRelation('foo', $foo);
            return $bar->foo ...; // Though safe for N+1 problem but ugly.
        },
    ]); ?>
</div>
```

``` php
<div class="bars-grid">
    <?= \yii\grid\GridView::widget([
        'dataProvider' => $dataProvider,
        'beforeRow' => function($bar) use($foo) {
            /** @var ActiveRecord $bar */
            $bar->populateRelation('foo', $foo);
        },
        'columns' => [
            'bar.name',
            'bar.foo.name',
            'bar.foo.other_property',
        ],
    ]); ?>
</div>
```

I want to move them to search model class from view template.

**Comment by fernandezekiel** [comment]
why not just eager load those models in the dataprovider's query object?

**Comment by tanakahisateru** [comment]
just In my case, eager loading can reduce SQL, but what I wanted is the SAME object of $foo. The first $foo already has various cached data, otherwise eagerly loaded ->foo is not and should do them again.

Yii can assign backward relation via inversedOf method. It supports to recycle first object.

But generally, this issue indicates that no one can assign model's AFTER_FIND handler dynamically in caller context. That event can be assigned only static context. One of Yii's advantage is dynamic event/behavior assigning in caller context instead of method overriding.

**Comment by dynasource** [comment]
related to 7268

**Comment by dynasource** [comment]
and #7187

**Comment by samdark** [comment]
Won't be implemented in 2.0.
